images/kindnetd/cmd/kindnetd/routes.go

```diff
@@ -24,30 +24,49 @@ import (
 	"k8s.io/klog/v2"
 )
 
-func syncRoute(nodeIP string, podCIDRs []string) error {
+func syncRoutes(nodeIP string, podCIDRs []string) error {
 	ip := net.ParseIP(nodeIP)
-
 	for _, podCIDR := range podCIDRs {
-		// parse subnet
-		dst, err := netlink.ParseIPNet(podCIDR)
-		if err != nil {
+		if err := syncRoute(podCIDR, ip); err != nil {
 			return err
 		}
+	}
+	return nil
+}
 
-		// Check if the route exists to the other node's PodCIDR
-		routeToDst := netlink.Route{Dst: dst, Gw: ip}
-		route, err := netlink.RouteListFiltered(nl.GetIPFamily(ip), &routeToDst, netlink.RT_FILTER_DST)
-		if err != nil {
-			return err
+func syncRoute(podCIDR string, gw net.IP) error {
+	// parse subnet
+	dst, err := netlink.ParseIPNet(podCIDR)
+	if err != nil {
+		return err
+	}
+
+	// List all routes to the other node's PodCIDR
+	routeToDst := netlink.Route{Dst: dst, Gw: gw}
+	routes, err := netlink.RouteListFiltered(nl.GetIPFamily(gw), &routeToDst, netlink.RT_FILTER_DST)
```

Comment:
does this list the default route to? the 0.0.0.0 ?

Reply:
puff, I'm asleep today, forget the comment, this is the same as before

Reply:
hmm, but this will only list the routes that match `dst`and `gw`, you want to obtain only the routes that match destination and all the gateways

Reply:
The `RouteListFiltered` is called with the mask `netlink.RT_FILTER_DST` which results in returning all routes which have the same `dst` as the given `routeToDst`. All other values of the Route are ignored during filtering.

Reply:
I see, so it was wrong before

Reply:
Example:

Consider the following active routing information:
```
$ ip -4 r
default via 172.18.0.1 dev eth0
172.18.0.0/16 dev eth0 proto kernel scope link src 172.18.0.6
192.168.0.0/24 via 172.18.0.7 dev eth0
192.168.1.0/24 via 172.18.0.5 dev eth0
192.168.3.0/24 via 172.18.0.7 dev eth0
192.168.4.0/24 via 172.18.0.9 dev eth0
192.168.5.0/24 via 172.18.0.5 dev eth0
192.168.6.0/24 via 172.18.0.4 dev eth0
192.168.8.0/24 via 172.18.0.9 dev eth0
```

* `syncRoutes("172.18.0.9", []string{"192.168.6.0/24"})`:
  * `netlink.RouteListFiltered` returns the following element: `192.168.6.0/24 via 172.18.0.4 dev eth0`
  * old implementation: would have skipped adding the new route which should look like `192.168.6.0/24 via 172.18.0.9 dev eth0`, if it even tried adding a new route, it would not have worked because there is already a route to `192.168.6.0/24` (error: `RTNETLINK answers: File exists`)
  * The new implementation removes `192.168.6.0/24 via 172.18.0.4 dev eth0` first and adds the new route instead

As visible in the example output above: there may still be orphaned routes, but they get recycled if the gateway IP changes.

Reply:
yes, yes, I'm a bit slow today, I see what you mean and I agree this reconciles it correctly, is just we never care about the case routes changed 🙃